lib/Differentiator/CladUtils.cpp

```diff
@@ -411,5 +411,17 @@ namespace clad {
     bool IsRValue(const clang::Expr* E) {
       return E->isRValue() || E->isXValue();
     }
+
+    clang::DeclContext* GetInnermostNamespaceOrTUnitDC(clang::Sema& semaRef,
```

Comment:
```suggestion
    clang::DeclContext* GetEnclosingNamespaceOrTUnitDC(clang::Sema& semaRef,
```

Can't we use [DeclContext::getEnclosingNamespaceContext](https://clang.llvm.org/doxygen/classclang_1_1DeclContext.html#aa838660af381c6967722c6552580c2a2)?

Reply:
We can, I totally forgot about this function. I will modify to code to use this function instead.

Reply:
I have slightly modified the custom derivative for class function design such that we no longer require this function. 

Now custom derivatives for class functions are always searched in the namespace: `clad::custom_derivatives::class_functions`. It does not take into account the namespace in which the class is defined. 
We can safely ignore the namespace in which class is defined because for each instance member function the first argument is a pointer to a class object that encodes class namespace information as well. For example: 

```cpp
namespace class_functions {

template <typename T>
void clear_pushforward(::std::vector<T>* v, ::std::vector<T>* d_v) {
  d_v->clear();
  v->clear();
}
```
The main motivation behind this change is that in libc++ vector is defined in `::std::__1` namespace, and in libstdc++ vector is defined in `::std` namespace. 

Please give your suggestion on this change.

Reply:
Yes, that makes sense to me. Why do we synthesize a pointer (eg `::std::vector<T>* v`) rather than a reference (`::std::vector<T>& v`)?

Reply:
It is to be more consistent with how C++ passes objects implicitly to member functions in the form of `this` pointer. 

We can easily change the requirements to pass by reference as well.

Reply:
We can make this change outside of this PR. I'd think it looks more natural to use `.` operators. That'd be expert feature but still probably better.

Reply:
In that case, should we ideally pass derivative object by reference as well?

Reply:
Absolutely.

Reply:
In that case, I am concerned that users would intuitively assume incorrect type of derivative functions. Since `this` is commonly assumed to be a pointer, people may expect derivative of `this` to be a pointer as well.

Reply:
Can you provide an example for it?

Reply:
Yes. 
Please consider the following code: 
```cpp
struct Complex {
  double real, im;
  double product() {
    return real*im;
  }
};
```
The member function `double product()` can also be written as: 

```cpp
double product() {
  return this->real * this->im;
}
```

Intuitively, the object is passed to a member function as an argument `this` -- that is of type pointer to the corresponding class of the member function. Therefore, since we are following that the derivative type is the same as of primal type, users may intuitively assume that the derivative of `this` argument is of the same type as of `this` as well -- pointer to the class.

Reply:
Thanks for the example. I don't think that should be a problem. I guess the same arguments hold for defining a copy constructor and yet people pass by reference.